pkg/util/buffered_chan.go

```diff
@@ -0,0 +1,91 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2021-present Datadog, Inc.
+
+package util
+
+import (
+	"context"
+	"sync"
+)
+
+// BufferedChan behaves like a `chan []interface{}` (See thread safety for restrictions), but is
+// most efficient as it uses internally a channel of []interface{}.
+// Thread safety:
+// 	- BufferedChan.Put cannot be called concurrently.
+// 	- BufferedChan.Get cannot be called concurrently.
+//  - BufferedChan.Put can be called while another gouroutine call BufferedChan.Get.
+type BufferedChan struct {
+	c        chan []interface{}
+	pool     *sync.Pool
+	putSlice []interface{}
+	getSlice []interface{}
+	getIndex int
+	ctx      context.Context
+}
+
+// NewBufferedChan creates a new instance of `BufferedChan`.
+// `ctx`` cancels all Put and Get operations.
+func NewBufferedChan(ctx context.Context, chanSize int, bufferSize int) *BufferedChan {
+	pool := &sync.Pool{
+		New: func() interface{} {
+			return make([]interface{}, 0, bufferSize)
+		},
+	}
+	return &BufferedChan{
+		c:        make(chan []interface{}, chanSize),
+		pool:     pool,
+		putSlice: pool.Get().([]interface{}),
+		ctx:      ctx,
+	}
+}
+
+// Put puts a new value into c.
+// Cannot be called concurrently.
+func (c *BufferedChan) Put(value interface{}) {
+	if cap(c.putSlice) <= len(c.putSlice) {
+		select {
+		case c.c <- c.putSlice:
+		case <-c.ctx.Done():
+		}
+		c.putSlice = c.pool.Get().([]interface{})[:0]
+	}
+	c.putSlice = append(c.putSlice, value)
+}
+
+// Close flushes and closes the channel
+func (c *BufferedChan) Close() {
+	if len(c.putSlice) > 0 {
+		c.c <- c.putSlice
+	}
+	close(c.c)
+}
+
+// Get gets the value and returns false when the channel is closed and all
+//  values were read.
+// Cannot be called concurrently.
+func (c *BufferedChan) Get() (interface{}, bool) {
+	if c.getIndex >= len(c.getSlice) {
+		if c.getSlice != nil {
+			c.pool.Put(c.getSlice)
```

Comment:
I guess there's no easy way to clean the content of slice that's put back (aka we have to wait for the old slice to be overwritten) to reduce memory pressure?

Reply:
I have added `c.getSlice[c.getIndex] = nil` just after `value := c.getSlice[c.getIndex]` to remove the reference in  the slice to the underlying object.
Note: The slice size is quite small and should not create any memory pressure anyway.

Reply:
Why can't we set length to zero before putting the slice into the pool?

Reply:
We can set the length of the slice to zero but I prefer to set the length to zero before using it (to make it clear the slice is empty).
I can also set the length of the slice to zero here if you want (I don't have a strong opinion on it). What do you think?

Reply:
I didn't see your previous comment yesterday for some reason. My conceren was that we may keep Serie longer than necessary, but if you remove references it should be fine.

I don't have a preference either way, but I think slices in the Pool should all be initialized the same. If you don't empty it on `Put`, then maybe make `New` return it the same way?

Reply:
AIUI _all_ elements in an array backing a slice are considered "in use" by the GC, so nothing is freed by `someSlice = someSlice[:0]`).  So what @ogaca-dd is doing (setting elements to `nil`) is the better choice for reducing memory pressure.  I thought I'd mention it to be explicit and to verify my understanding.